kubeflow/tf-serving/prototypes/tf-serving-all-features.jsonnet

```diff
@@ -8,6 +8,12 @@
 // @optionalParam model_server_image string gcr.io/kubeflow-images-staging/tf-model-server:v20180227-master Container for TF model server
 // @optionalParam http_proxy_image string gcr.io/kubeflow/http-proxy:1.0 Container for http_proxy of TF model server
 // @optionalParam service_type string ClusterIP The service type for TFServing deployment.
+// @optionalParam aws_access_key_id string key_id S3 access key id
+// @optionalParam aws_secret_access_key string secret_key S3 secret access key
```

Comment:
Is the access_key something you would want to check into source control?

Reply:
That's up to the user. However they store it, it can be passed in as a param to `ks generate`

Reply:
ksonnet apps are meant to be checked into source control. This means users should generally not store sensitive information in the ksonnet app e.g passwords.

So if aws_secret_access_key is a password then it shouldn't be a parameter because that would mean it gets checked into source control.

In this case a better approach would be to take as a parameter the name of a K8s secret which stores the password.

Reply:
well this ks will generate secret out of it too, not sure which is better - require secret pre-created or create from params here? In helm it would be the latter I guess.

Reply:
From a user perspective it never even occurred to me that you were supposed to check the output of ksonnet in somewhere. In my case I'm using ksonnet within a container in an argo workflow to deploy a model for serving. The ksonnet application just evaporates leaving the model running in k8s.

I also think it's better UX to be able to provide credentials via ks. This way you avoid having someone mess with kubectl directly in the middle of their workflow (whether automated or not)

As far as checking secrets in to git, I think it's fine as long as something like git-crypt is used. I can add a disclaimer advising such protections for those wanting to check in the output of ks.

Reply:
Whether one is using ksonnet or raw yaml, I think one of the points of describing infrastructure decoratively is so that you can check it into version control. 

So I really don't think we should structure our configs in a way that would lead to secrets being stored in version control.

I think the long term solution here is that K8s will support external key management systems. So that you can tell K8s to create a secret from some key in your KMS and as part of deploying the infrastructure K8s will check whether you have permission to access the key and fetch it. T

The point is our ksonnet/yaml configs should never contain any data that is sensitive and should be encrypted; only unsensitive references (e.g. the name of a key, name of a file etc)...

So I don't think we should put secrets into our ksonnet configs.

I think the general pattern that we should encourage users to follow is you create a secret outside of ksonnet and then pass the name of the secret as a parameter.

An example is our test infrastructure
https://github.com/kubeflow/testing#create-a-gcp-service-account